### PR TITLE
util: make list.MemoryUsage() be more efficient

### LINKED ISSUE #5722: list.MemoryUsage is inefficient
list.go
```go
129 // MemoryUsage calculates the total memory usage of a list.
130 func (l *List) MemoryUsage() (sum int64) {
131     for _, c := range l.chunks {
132         sum += c.MemoryUsage()
133     }
134     for _, f := range l.freelist {
135         sum += f.MemoryUsage()
136     }
137     return sum
138 }
```

join.go
```go
 355         innerMemUsage := innerResult.MemoryUsage()
 356         if !memExceedThreshold && innerMemUsage > execMemThreshold {
 357             memExceedThreshold = true
 358             log.Warnf(ErrMemExceedThreshold.GenByArgs("HashJoin", innerMemUsage, execMemThreshold).Error())
 359         }
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
fix #5722
After this commit, we only calculate the memory usage of `list.chunks[len(list.chunks)-1]`
when call list.MemoryUsage() every time.

PTAL @zz-jason

### CODE REVIEW

**Comment by zz-jason** [comment]
please add some bench test about `List.MemUsage()` and `Chunk.MemUsage()`

**Comment by XuHuaiyu** [comment]
Before:
BenchmarkListMemoryUsage-4       5000000               337 ns/op
After:
BenchmarkListMemoryUsage-4      100000000               18.1 ns/op

File: util/chunk/list_test.go [hunk 1]
```diff
@@ -17,6 +17,10 @@ import (
 	"github.com/pingcap/check"
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
+	"math"
```
Comment (coocood) [inline util/chunk/list_test.go]: Please fix the import order.

File: util/chunk/list_test.go [hunk 2]
```diff
@@ -17,6 +17,10 @@ import (
...
+	"testing"
+	"time"
```
Comment (lamxTyler) [inline util/chunk/list_test.go]: Put these standard package upward.

**Comment by shenli** [comment]
@lamxTyler @coocood PTAL

File: util/chunk/list.go [hunk 3]
```diff
@@ -25,6 +25,9 @@ type List struct {
 	length       int
 	chunks       []*Chunk
 	freelist     []*Chunk
+	// memUsage[0] is the memory usage of `chunks`,
+	// memUsage[1] is the memory usage of `freelist`.
+	memUsage []int64
```
Comment (coocood) [inline util/chunk/list.go]: Why not just use two fields? `memUsageChunk` and `memUsageFreeList`?

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**